ethereum_history_api/circuits/lib/src/serialize.nr

```diff
@@ -0,0 +1,72 @@
+use crate::account::AccountWithinBlock;
+use crate::account_with_storage::{StorageWithinBlock, Account};
+use crate::misc::{fragment::Fragment, types::{BYTES32_LENGTH, Bytes32, ADDRESS_LENGTH, Address}};
+use dep::std::unsafe::zeroed;
+
+trait Serialize<LEN> {
+    fn serialize(self) -> [Field; LEN];
+}
+
+global U128_SERIALIZED_LEN = 2;
+
+impl Serialize<U128_SERIALIZED_LEN> for U128 {
+    fn serialize(self) -> [Field; U128_SERIALIZED_LEN] {
+        [self.lo, self.hi]
+    }
+}
+
+impl Serialize<BYTES32_LENGTH> for Bytes32 {
+    fn serialize(self) -> [Field; BYTES32_LENGTH] {
+        let mut data: [Field; BYTES32_LENGTH] = zeroed();
+        for i in 0..BYTES32_LENGTH {
+            data[i] = self[i] as Field;
+        }
+        data
+    }
+}
+
+impl Serialize<ADDRESS_LENGTH> for Address {
+    fn serialize(self) -> [Field; ADDRESS_LENGTH] {
+        let mut data: [Field; ADDRESS_LENGTH] = zeroed();
+        for i in 0..ADDRESS_LENGTH {
+            data[i] = self[i] as Field;
+        }
+        data
+    }
+}
+
+global ACCOUNT_SERIALIZED_LEN = 1 + 1 + BYTES32_LENGTH + BYTES32_LENGTH;
+
+impl Serialize<ACCOUNT_SERIALIZED_LEN> for Account {
+    fn serialize(self) -> [Field; ACCOUNT_SERIALIZED_LEN] {
+        let mut data: Fragment<ACCOUNT_SERIALIZED_LEN, Field> = Fragment::empty();
+        data.push_back(self.nonce as Field);
+        data.push_back(self.balance);
+        data.extend_back(self.storage_root.serialize());
+        data.extend_back(self.code_hash.serialize());
+        data.to_array()
+    }
+}
+
+global ACCOUNT_WITHIN_BLOCK_SERIALIZED_LEN = ACCOUNT_SERIALIZED_LEN + BYTES32_LENGTH;
```

Comment:
Does it mean adding globals to use in generics now works?

Reply:
Yes. But arithmetic on generics still does not